composer.json

```diff
@@ -32,6 +32,8 @@
         "composer/xdebug-handler": "^1.4|^2.0",
         "danielstjules/stringy": "^3.1",
         "doctrine/inflector": "^2.0",
+        "ergebnis/json-printer": "^3.1",
```

Comment:
Instead we already use `Symplify\SmartFileSystem\Json\JsonFileSystem` that does the same job. Could you change it?

Reply:
@TomasVotruba No, the JsonFileSystem does only a json_encode behind the scenes. And this is not really configurable hence this package exists. json_encode uses 4 spaces to print the file. Or do i miss something here?

Reply:
So does the JsonFileSystem. We should use just once package instead of 2 :)

Reply:
I think we have a misunderstanding. The ergebnis/json-printer package is highly configurable json printer and does more than a json_encode: https://localheinz.com/blog/2018/01/04/pretty-printing-json-with-custom-indent/

Reply:
I see :+1: 

Do we need per suffix formatters? The space and indents are the same everywhere.

Reply:
Could you add formatter for e.g. `*.twig` file? It should be more obvious then

Reply:
> I see 👍
> 
> Why do we need per suffix formatters? The space and indents are the same everywhere.
> IMO there should be one generic formatter that works with any suffix.

I don´t know if this is possible. I guess you always need some dedicated specialized formatters for xml, json, html etc. I fear there is no one fits all solutions.

Reply:
> Could you add formatter for e.g. `*.twig` file? It should be more obvious then

I am not aware of any twig formatter. What would you expect here?

Reply:
Same test like this, just for twig:
https://github.com/rectorphp/rector/pull/6272/files#diff-0d87c4ff1a85d1f7b5b59e429c631725f403cd74d7f0852fd13798d00704c9ac

> I don´t know if this is possible. I guess you always need some dedicated specialized formatters for xml, json, html etc. I fear there is no one fits all solutions.

To make this feature complete we need to cover those we support already:

- json
- neon
- yaml
- twig
- latte

Reply:
Yaml and neon should no be  a problem, but latte and twig. I am not sure.
The NonPhpFileProcessor is not really writing back a whole but only replacing some strings as far as i can see. 
So, no need to have a real formatter here.

Reply:
I see. What format do we have to cover then?

Reply:
I would say, xml, yaml, neon should be fine to start with.

Reply:
I know it is some maintenance burden, if you don´t like the idea or feel resistance, we can still drop it.

Reply:
I'm just trying to understand what needs to done to make it work for existing processors.

> I would say, xml, yaml, neon should be fine to start with.

OK. Could you add those?

Reply:
Neon is harder than i expected. I thought that there is also a decoder with format options. But there isn´t. 

Maybe this whole formatting is a bad idea. It is really tough to cover all formats. In my case it is not only about changing tabs to spaces or indentation from 2 to 4. E.g. i am changing xml files and adding nodes, but preserve formatting is not that easy after modifying the xml structure.

Reply:
I see. Lets stick what minimum you need at the moment. We'll extend it if needed in the future.

Let me know when it's ready to merge :+1: